crates/wasi/src/preview2/tcp.rs

```diff
@@ -46,113 +45,206 @@ pub(crate) enum HostTcpState {
+                // primitive try_write, which goes directly to attempt a write with mio. This has
+                // two advantages: 1. this operation takes a &TcpStream instead of a &mut TcpStream
+                // required to AsyncWrite, and 2. it eliminates any buffering in tokio we may need
+                // to flush.
+                while !bytes.is_empty() {
+                    stream.writable().await?;
+                    let n = stream.try_write(&bytes)?;
+                    let _ = bytes.split_to(n);
+                }
 
-        tcp_socket
+                Ok(())
+            })
+            .into(),
+        );
     }
 }
 
 #[async_trait::async_trait]
-impl HostInputStream for Arc<HostTcpSocketInner> {
-    fn read(&mut self, size: usize) -> anyhow::Result<(Bytes, StreamState)> {
-        if size == 0 {
-            return Ok((Bytes::new(), StreamState::Open));
+impl HostOutputStream for TcpWriteStream {
+    fn write(&mut self, mut bytes: bytes::Bytes) -> Result<(), OutputStreamError> {
+        while !bytes.is_empty() {
+            match self.stream.try_write(&bytes) {
+                Ok(n) => {
+                    let _ = bytes.split_to(n);
+                }
+
+                Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                    // As `try_write` indicated that it would have blocked, we'll perform the write
+                    // in the background to allow us to return immediately.
+                    self.background_write(bytes);
+
+                    return Ok(());
+                }
+
+                Err(e) => return Err(OutputStreamError::LastOperationFailed(e.into())),
+            }
         }
-        let mut buf = BytesMut::zeroed(size);
-        let r = self
-            .tcp_socket()
-            .as_socketlike_view::<TcpStream>()
-            .read(&mut buf);
-        let (n, state) = read_result(r)?;
-        buf.truncate(n);
-        Ok((buf.freeze(), state))
+
+        Ok(())
     }
 
-    async fn ready(&mut self) -> anyhow::Result<()> {
-        self.tcp_socket.readable().await?;
+    fn flush(&mut self) -> Result<(), OutputStreamError> {
+        // `flush` is a no-op here, as we're not managing any internal buffer. Additionally,
+        // `write_ready` will join the background write task if it's active, so following `flush`
+        // with `write_ready` will have the desired effect.
         Ok(())
     }
-}
 
-#[async_trait::async_trait]
-impl HostOutputStream for Arc<HostTcpSocketInner> {
-    fn write(&mut self, buf: Bytes) -> anyhow::Result<(usize, StreamState)> {
-        if buf.is_empty() {
-            return Ok((0, StreamState::Open));
+    async fn write_ready(&mut self) -> Result<usize, OutputStreamError> {
+        if self.write_handle.is_some() {
+            self.write_handle
+                .as_mut()
+                .unwrap()
+                .await
+                .map_err(|e| OutputStreamError::Trap(e.into()))?
```

Comment:
Like way up a above IMO this should be `.unwrap` or `.expect` to propagate a panic rather than hiding it under a trap

Reply:
i changed AbortOnDropJoinHandle's Future impl to always unwrap the task panics